[HOLESKY PECTRA] `ArithmeticException`! FCU

```
{"@timestamp":"2025-03-02T20:53:40,855","level":"ERROR","thread":"forkchoice-async-0","class":"ForkChoiceRatchet","message":"Fork choice process head failed","throwable":" java.util.concurrent.CompletionException: java.lang.ArithmeticException: uint64 overflow\n\tat java.base/java.util.concurrent.CompletableFuture.encodeRelay(CompletableFuture.java:368)\n\tat java.base/java.util.concurrent.CompletableFuture.completeRelay(CompletableFuture.java:377)\n\tat java.base/java.util.concurrent.CompletableFuture$UniRelay.tryFire(CompletableFuture.java:1097)\n\tat java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)\n\tat java.base/java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:2194)\n\tat tech.pegasys.teku.infrastructure.async.SafeFuture.lambda$propagateResult$3(SafeFuture.java:145)\n\tat java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:863)\n\tat java.base/java.util.concurrent.CompletableFuture.uniWhenCompleteStage(CompletableFuture.java:887)\n\tat java.base/java.util.concurrent.CompletableFuture.whenComplete(CompletableFuture.java:2357)\n\tat tech.pegasys.teku.infrastructure.async.SafeFuture.whenComplete(SafeFuture.java:620)\n\tat tech.pegasys.teku.infrastructure.async.SafeFuture.whenComplete(SafeFuture.java:31)\n\tat tech.pegasys.teku.infrastructure.async.SafeFuture.propagateResult(SafeFuture.java:142)\n\tat tech.pegasys.teku.infrastructure.async.SafeFuture.propagateTo(SafeFuture.java:318)\n\tat tech.pegasys.teku.infrastructure.async.ScheduledExecutorAsyncRunner.lambda$createRunnableForAction$1(ScheduledExecutorAsyncRunner.java:124)\n\tat java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)\n\tat java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)\n\tat java.base/java.lang.Thread.run(Thread.java:1583)\nCaused by: java.lang.ArithmeticException: uint64 overflow\n\tat tech.pegasys.teku.infrastructure.unsigned.UInt64.plus(UInt64.java:142)\n\tat tech.pegasys.teku.infrastructure.unsigned.UInt64.plus(UInt64.java:108)\n\tat tech.pegasys.teku.storage.protoarray.ProtoNode.adjustWeight(ProtoNode.java:108)\n\tat tech.pegasys.teku.storage.protoarray.ProtoArray.applyDelta(ProtoArray.java:728)\n\tat tech.pegasys.teku.storage.protoarray.ProtoArray.lambda$applyDeltas$8(ProtoArray.java:716)\n\tat tech.pegasys.teku.storage.protoarray.ProtoArray.applyToNodes(ProtoArray.java:744)\n\tat tech.pegasys.teku.storage.protoarray.ProtoArray.applyDeltas(ProtoArray.java:716)\n\tat tech.pegasys.teku.storage.protoarray.ProtoArray.applyScoreChanges(ProtoArray.java:429)\n\tat tech.pegasys.teku.storage.protoarray.ForkChoiceStrategy.applyPendingVotes(ForkChoiceStrategy.java:126)\n\tat tech.pegasys.teku.storage.store.StoreVoteUpdater.applyForkChoiceScoreChanges(StoreVoteUpdater.java:81)\n\tat tech.pegasys.teku.statetransition.forkchoice.ForkChoice.updateHeadTransaction(ForkChoice.java:380)\n\tat tech.pegasys.teku.statetransition.forkchoice.ForkChoice.lambda$processHead$11(ForkChoice.java:353)\n\tat tech.pegasys.teku.infrastructure.async.eventthread.AsyncRunnerEventThread.recordEventThreadIdAndExecute(AsyncRunnerEventThread.java:130)\n\tat tech.pegasys.teku.infrastructure.async.eventthread.AsyncRunnerEventThread.lambda$doExecute$1(AsyncRunnerEventThread.java:113)\n\tat tech.pegasys.teku.infrastructure.async.SafeFuture.of(SafeFuture.java:80)\n\tat tech.pegasys.teku.infrastructure.async.AsyncRunner.lambda$runAsync$2(AsyncRunner.java:47)\n\tat tech.pegasys.teku.infrastructure.async.SafeFuture.of(SafeFuture.java:72)\n\t... 4 more\n"}
```

holesky-03

--- a/infrastructure/unsigned/src/main/java/tech/pegasys/teku/infrastructure/unsigned/UInt64.java
+++ b/infrastructure/unsigned/src/main/java/tech/pegasys/teku/infrastructure/unsigned/UInt64.java
@@ -30,6 +30,7 @@ public final class UInt64 implements Comparable<UInt64> {
   public static final UInt64 ONE = new UInt64(1);
   public static final UInt64 THIRTY_TWO_ETH = new UInt64(32_000_000_000L);
   public static final UInt64 MAX_VALUE = new UInt64(-1L);
+  public static final UInt64 SAFE_MAX_VALUE = new UInt64(Long.MAX_VALUE - 1);
 
   private final long value;
 
@@ -182,6 +183,19 @@ public final class UInt64 implements Comparable<UInt64> {
     }
 
     return Optional.of(fromLongBits(value - other));
+  }
+
+  public Optional<UInt64> safePlus(final long other) {
+    checkPositive(other);
+    if (UInt64.SAFE_MAX_VALUE.minus(other).isGreaterThanOrEqualTo(value)) {
+      return Optional.of(fromLongBits(value + other));
+    }
+    return Optional.empty();
+  }
+
+  public Optional<UInt64> safePlus(final UInt64 other) {
+    checkPositive(other.value);
+    return safePlus(other.value);
   }
 
   /**

--- a/infrastructure/unsigned/src/main/java/tech/pegasys/teku/infrastructure/unsigned/UInt64.java
+++ b/infrastructure/unsigned/src/main/java/tech/pegasys/teku/infrastructure/unsigned/UInt64.java
@@ -30,7 +30,6 @@ public final class UInt64 implements Comparable<UInt64> {
   public static final UInt64 ONE = new UInt64(1);
   public static final UInt64 THIRTY_TWO_ETH = new UInt64(32_000_000_000L);
   public static final UInt64 MAX_VALUE = new UInt64(-1L);
-  public static final UInt64 SAFE_MAX_VALUE = new UInt64(Long.MAX_VALUE - 1);
 
   private final long value;
 
@@ -145,6 +144,19 @@ public final class UInt64 implements Comparable<UInt64> {
     return fromLongBits(longBits1 + longBits2);
   }
 
+  public UInt64 safePlus(final long other) {
+    checkPositive(other);
+    if (value != 0 && Long.compareUnsigned(other, MAX_VALUE.longValue() - value) > 0) {
+      return UInt64.MAX_VALUE;
+    }
+    return plus(value, other);
+  }
+
+  public UInt64 safePlus(final UInt64 other) {
+    checkPositive(other.value);
+    return safePlus(other.value);
+  }
+
   /**
    * Return the result of subtracting the specified value from this one.
    *
@@ -183,19 +195,6 @@ public final class UInt64 implements Comparable<UInt64> {
     }
 
     return Optional.of(fromLongBits(value - other));
-  }
-
-  public Optional<UInt64> safePlus(final long other) {
-    checkPositive(other);
-    if (UInt64.SAFE_MAX_VALUE.minus(other).isGreaterThanOrEqualTo(value)) {
-      return Optional.of(fromLongBits(value + other));
-    }
-    return Optional.empty();
-  }
-
-  public Optional<UInt64> safePlus(final UInt64 other) {
-    checkPositive(other.value);
-    return safePlus(other.value);
   }
 
   /**

--- a/infrastructure/unsigned/src/test/java/tech/pegasys/teku/infrastructure/unsigned/UInt64Test.java
+++ b/infrastructure/unsigned/src/test/java/tech/pegasys/teku/infrastructure/unsigned/UInt64Test.java
@@ -19,7 +19,6 @@ import static org.assertj.core.api.Assertions.assertThatThrownBy;
 
 import java.math.BigInteger;
 import java.util.List;
-import java.util.Optional;
 import java.util.stream.IntStream;
 import java.util.stream.Stream;
 import org.junit.jupiter.api.Test;
@@ -320,7 +319,7 @@ class UInt64Test {
   @ParameterizedTest
   @MethodSource("safePlusNumbers")
   void safePlus_shouldAddWhenNotOverflowingLongs(
-      final long value1, final long value2, final Optional<UInt64> expected) {
+      final long value1, final long value2, final UInt64 expected) {
     final UInt64 uint1 = UInt64.fromLongBits(value1);
     assertThat(uint1.safePlus(value2)).isEqualTo(expected);
   }
@@ -641,13 +640,13 @@ class UInt64Test {
   }
 
   static Stream<Arguments> safePlusNumbers() {
-    final long max = UInt64.SAFE_MAX_VALUE.longValue();
+    final long max = UInt64.MAX_VALUE.longValue();
     return Stream.of(
-        Arguments.arguments(0L, max, Optional.of(UInt64.SAFE_MAX_VALUE)),
-        Arguments.arguments(max, 0L, Optional.of(UInt64.SAFE_MAX_VALUE)),
-        Arguments.arguments(max - 10L, 10L, Optional.of(UInt64.SAFE_MAX_VALUE)),
-        Arguments.arguments(1L, 10L, Optional.of(UInt64.valueOf(11))),
-        Arguments.arguments(max, 1L, Optional.empty()));
+        Arguments.arguments(max, 0L, UInt64.MAX_VALUE),
+        Arguments.arguments(max - 10L, 10L, UInt64.MAX_VALUE),
+        Arguments.arguments(max - 11L, 10L, UInt64.MAX_VALUE.minus(1)),
+        Arguments.arguments(1L, 10L, UInt64.valueOf(11)),
+        Arguments.arguments(max, 1L, UInt64.MAX_VALUE));
   }
 
   static List<Arguments> rangeNumbers() {

--- a/infrastructure/unsigned/src/test/java/tech/pegasys/teku/infrastructure/unsigned/UInt64Test.java
+++ b/infrastructure/unsigned/src/test/java/tech/pegasys/teku/infrastructure/unsigned/UInt64Test.java
@@ -19,6 +19,7 @@ import static org.assertj.core.api.Assertions.assertThatThrownBy;
 
 import java.math.BigInteger;
 import java.util.List;
+import java.util.Optional;
 import java.util.stream.IntStream;
 import java.util.stream.Stream;
 import org.junit.jupiter.api.Test;
@@ -314,6 +315,14 @@ class UInt64Test {
         .isInstanceOf(ArithmeticException.class);
     assertThatThrownBy(() -> UInt64.MAX_VALUE.plus(UInt64.MAX_VALUE))
         .isInstanceOf(ArithmeticException.class);
+  }
+
+  @ParameterizedTest
+  @MethodSource("safePlusNumbers")
+  void safePlus_shouldAddWhenNotOverflowingLongs(
+      final long value1, final long value2, final Optional<UInt64> expected) {
+    final UInt64 uint1 = UInt64.fromLongBits(value1);
+    assertThat(uint1.safePlus(value2)).isEqualTo(expected);
   }
 
   @Test
@@ -629,6 +638,16 @@ class UInt64Test {
     assertThat(UInt64.range(UInt64.valueOf(from), UInt64.valueOf(to)))
         .containsExactlyElementsOf(
             IntStream.range(from, to).mapToObj(UInt64::valueOf).collect(toList()));
+  }
+
+  static Stream<Arguments> safePlusNumbers() {
+    final long max = UInt64.SAFE_MAX_VALUE.longValue();
+    return Stream.of(
+        Arguments.arguments(0L, max, Optional.of(UInt64.SAFE_MAX_VALUE)),
+        Arguments.arguments(max, 0L, Optional.of(UInt64.SAFE_MAX_VALUE)),
+        Arguments.arguments(max - 10L, 10L, Optional.of(UInt64.SAFE_MAX_VALUE)),
+        Arguments.arguments(1L, 10L, Optional.of(UInt64.valueOf(11))),
+        Arguments.arguments(max, 1L, Optional.empty()));
   }
 
   static List<Arguments> rangeNumbers() {

--- a/storage/src/main/java/tech/pegasys/teku/storage/protoarray/ProtoNode.java
+++ b/storage/src/main/java/tech/pegasys/teku/storage/protoarray/ProtoNode.java
@@ -109,14 +109,10 @@ public class ProtoNode {
       }
       weight = weight.minus(deltaAbsoluteValue);
     } else {
-      weight
-          .safePlus(delta)
-          .ifPresentOrElse(
-              newValue -> weight = newValue,
-              () -> {
-                LOG.trace("Unable to add delta {} to weight {}", delta, weight);
-                weight = UInt64.SAFE_MAX_VALUE;
-              });
+      weight = weight.safePlus(delta);
+      if (weight.equals(UInt64.MAX_VALUE)) {
+        LOG.trace("Unable to add delta {} to weight {}", delta, weight);
+      }
     }
   }
 

--- a/storage/src/main/java/tech/pegasys/teku/storage/protoarray/ProtoNode.java
+++ b/storage/src/main/java/tech/pegasys/teku/storage/protoarray/ProtoNode.java
@@ -18,6 +18,8 @@ import static com.google.common.base.Preconditions.checkState;
 import com.google.common.base.MoreObjects;
 import java.util.Objects;
 import java.util.Optional;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
 import org.apache.tuweni.bytes.Bytes32;
 import tech.pegasys.teku.infrastructure.logging.LogFormatter;
 import tech.pegasys.teku.infrastructure.unsigned.UInt64;
@@ -30,6 +32,8 @@ public class ProtoNode {
 
   public static final UInt64 NO_EXECUTION_BLOCK_NUMBER = UInt64.ZERO;
   public static final Bytes32 NO_EXECUTION_BLOCK_HASH = Bytes32.ZERO;
+
+  private static final Logger LOG = LogManager.getLogger();
 
   private final UInt64 blockSlot;
   private final Bytes32 stateRoot;
@@ -105,7 +109,12 @@ public class ProtoNode {
       }
       weight = weight.minus(deltaAbsoluteValue);
     } else {
-      weight = weight.plus(delta);
+      if (UInt64.MAX_VALUE.minusMinZero(delta).isGreaterThan(weight)) {
+        weight = weight.plus(delta);
+      } else {
+        LOG.trace("Unable to add delta {} to weight {}", delta, weight);
+        weight = UInt64.MAX_VALUE;
+      }
     }
   }
 

--- a/storage/src/main/java/tech/pegasys/teku/storage/protoarray/ProtoNode.java
+++ b/storage/src/main/java/tech/pegasys/teku/storage/protoarray/ProtoNode.java
@@ -109,12 +109,14 @@ public class ProtoNode {
       }
       weight = weight.minus(deltaAbsoluteValue);
     } else {
-      if (UInt64.MAX_VALUE.minusMinZero(delta).isGreaterThan(weight)) {
-        weight = weight.plus(delta);
-      } else {
-        LOG.trace("Unable to add delta {} to weight {}", delta, weight);
-        weight = UInt64.MAX_VALUE;
-      }
+      weight
+          .safePlus(delta)
+          .ifPresentOrElse(
+              newValue -> weight = newValue,
+              () -> {
+                LOG.trace("Unable to add delta {} to weight {}", delta, weight);
+                weight = UInt64.SAFE_MAX_VALUE;
+              });
     }
   }
 

